plugins/secom_cl.py

```diff
@@ -196,322 +221,463 @@ def sem_roi_to_ccd(escan, ccd, roi, margin=0):
         """
-        Stop spot mode
+        TODO
+        :param future: ()
         """
-        # unsubscribe to the data, it will automatically stop the spot
-        self.edet.data.unsubscribe(self._receive_sem_data)
-        self.edet.data.synchronizedOn(None)
+        self.ccd_roi = sem_roi_to_ccd(self._emitter, self._ccd, self.roi.value, self._sccd.roi_margin.value)
+
+        return super(SECOMCLSEMMDStream, self)._runAcquisition(future)
 
-    def _receive_sem_data(self, df, data):
+    def _preprocessData(self, n, data, i):
         """
-        Store SEM data (when scanning spot mode typically)
+        Preprocess the raw data, just after it was received from the detector.
+        :param n: (0<=int) The detector/stream index.
+        :param data: (DataArray) The data as received from the detector, from
+        _onData(), and with MD_POS updated to the current position of the e-beam.
+        :param i: (int, int) The iteration number in X, Y.
+        :returns: (value) The value as needed by _onCompletedData.
         """
-        self._sem_data.append(data)
-        self._sem_data_received.set()
-        if data.shape != (1,1):
-            logging.warning("SEM data shape is %s while expected a spot", data.shape)
+        if n != self._ccd_idx:
+            return super(SECOMCLSEMMDStream, self)._preprocessData(n, data, i)
 
-    def stop_acquisition(self):
-        self._must_stop = True
-        logging.debug("Cancelling acquisition")
+        ccd_roi = self.ccd_roi
+        data = data[ccd_roi[1]: ccd_roi[3] + 1, ccd_roi[0]: ccd_roi[2] + 1]  # crop
 
-    def start_ccd(self):
-        self.ccd.data.synchronizedOn(self.ccd.softwareTrigger)
-        self.ccd.data.subscribe(self._receive_ccd_data)
+        # memu = acqstream.udriver.readMemoryUsage()
+        # # Too bad, need to use VmSize to get any good value
+        # logging.info("Memory usage is %d. ", memu)
 
-    def stop_ccd(self):
-        self.ccd.data.unsubscribe(self._receive_ccd_data)
-        self.ccd.data.synchronizedOn(None)
+        cpos = self._get_center_pos(data, self.ccd_roi)
+        sname = self._streams[n].name.value
 
-    def _receive_ccd_data(self, df, data):
-        """
-        Store CCD data
-        """
-        self._ccd_data.append(data)
-        self._ccd_data_received.set()
+        data.metadata[model.MD_DESCRIPTION] = sname
+        # update center position of optical image (should be the same for all optical images)
+        data.metadata[model.MD_POS] = cpos
 
-    def acquire_ccd(self, x, y, ccd_roi_idx):
-        """
-        Acquire an image from the CCD while having the e-beam at a spot position
-        x, y (floats): spot position in the ebeam coordinates
-        ccd_roi_idx: slice to crop the CCD image
-        return (model.DataArray of shape (Y,X), model.DataArray of shape (1,1)):
-          the CCD image and the SEM data (at the spot)
-        """
-        self.move_spot(x, y)
+        fn = self.filename.value
```

Comment:
Just before this part of the code, you need to put a BIG comment explaining this gigantic hack.